Create region sensor

dan-rusu-inc opened the pull request:
Created region sensors, fixed some defects in HashSet & UnorderedArrayList, plus some cleanup

daniel-rusu committed:
Merge remote-tracking branch 'refs/remotes/RoosterEngine/master'

daniel-rusu committed:
Create region sensor + cleanup

Created region sensors, fixed some defects with the HashSet &
UnorderedArrayList, and some cleanup.

daniel-rusu committed:
Complex input handling not ready to submit yet

Accidentally included these files in the previous commit but they're not
ready yet

davidrusu commented on src/Utilities/HashSet.java:
```diff
@@ -1,15 +1,19 @@
  * @author davidrusu
  */
-public class HashSet {
+public class HashSet<T> {
```
👍

davidrusu commented on src/gameengine/collisiondetection/tree/Tree.java:
```diff
@@ -133,17 +134,20 @@ public void entityUpdated(double timeToCheck, Entity entity) {
-            }
+        if ((world.getCollisionGroups()[a.getEntityType()] & b.getEntityTypeBitMask()) == 0 || a
+                instanceof RegionSensor && ((RegionSensor) a).containsEntity(b) || b instanceof
+                RegionSensor && ((RegionSensor) b).containsEntity(a)) {
```
What happens if both `a` and `b` are both region sensors?

dan-rusu-inc replied:
So if the first condition (& bitmask) == 0 then we know that a and b don't interact and return right away.  If we keep checking the condition that means that the first condition is false so they should interact which means that the region sensors want to be aware of each other.  If the one of the sensors already contains the other then they are already aware.

But... This does point out a defect in a different part of the code where I add an entity to the region sensor as I should add each other to the other (whereas right now only 1 region sensor is  notified).

davidrusu commented on src/gameengine/context/Context.java:
```diff
@@ -259,7 +260,7 @@ private RepeatedAction(long repeatDelay, boolean zeroDelayToFirstAction, boolean
             assert repeatDelay > 0;
             this.action = action;
-            delay = repeatDelay * 1000;
+            delay = repeatDelay * GameTimer.NANOS_PER_MILLI;
```
👍

davidrusu commented on src/gameengine/input/MouseController.java:
```diff
@@ -91,17 +91,11 @@ public MouseController(GameCore core, ScreenManager screen, int desiredFramerate
 
-    /**
-     * @see MouseProperties#isInvertedVerticalAxis()
-     */
```
👍

davidrusu commented on src/gameengine/physics/Physics.java:
```diff
@@ -20,7 +20,10 @@ public static double getCriticallyDampedSpringConstant(double k, double mass) {
+        double dx = a.getX() - b.getX();
+        double dy = a.getY() - b.getY();
+        double length = Math.sqrt(dx * dx + dy * dy);
+        double displacement = length - restingLength;
```
😢

daniel-rusu committed:
Allow region sensors to be aware of other region sensors

Depending on the collision groups, region sensors could be configured to
be aware of other region sensors so if 2 region sensors collide then
they both need to be aware of each other.
